#ID0309 - Aprendizaje Estadístico

### Otoño 2023

### ID0309_Lab_1-3

**Enrique Naredo García**

<font size = 2>
©️ Todos los derechos reservados. All rights reserved.

*Nota: El presente documento es una herramienta diseñada única y exclusivamente para los estudiantes de la asignatura arriba mencionada. Se recuerda no compartir esta información fuera de los integrantes registrados en este curso. La reproducción total o parcial de este documento requiere autorización por escrito del titular del copyright.*
</font>

# Limpieza de Datos

Utiizaremos el conjunto de datos: "BL-Flickr-Images-Book", el cuál es un archivo "csv", que contiene información sobre libros de la Biblioteca Británica.

En este Jupyter notebook aprenderemos a realizar las siguientes actividades:

* Eliminar columnas.
* Cambiar el índice.
* Ordenar campos.
* Combinando métodos

##Eliminar columnas

A menudo, descubrirás que no todas las categorías de datos de un conjunto de datos le resultan útiles.

* Por ejemplo, es posible que tenga un conjunto de datos que contenga información de los estudiantes.
* Nombre, grado, estándar, nombres de los padres y dirección.
* Pero desee centrarse en analizar las calificaciones de los estudiantes.

En este caso, las categorías de dirección o nombre de los padres no son importantes.

* Retener estas categorías innecesarias ocupará espacio innecesario y potencialmente también atascará el tiempo de ejecución.

* Pandas proporciona una forma práctica de eliminar columnas o filas no deseadas de un DataFrame con la función drop().


In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('/content/drive/MyDrive/Datasets/BL-Flickr-Images-Book.csv')
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

Cuando miramos las primeras cinco entradas usando el método head().
* Podemos ver que algunas columnas proporcionan información auxiliar que sería útil para la biblioteca.
* Pero que no es muy descriptiva de los libros en sí:
  * Declaración de edición,
  * Autor corporativo,
  * Colaboradores corporativos,
  * propietario anterior,
  * grabador,
  * tipo de emisión, y
  * marcas comerciales.

Podemos eliminar estas columnas de la siguiente manera:
* Definimos una lista que contiene los nombres de todas las columnas que queremos eliminar.

In [4]:
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']

df.drop(to_drop, inplace=True, axis=1)

Observamos los datos después de este proceso.

In [5]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

##Cambiar el índice

Un índice de pandas amplía la funcionalidad de las matrices NumPy para permitir un corte y etiquetado más versátil.

* En muchos casos, resulta útil utilizar un campo de identificación de los datos con un valor único como índice.

* Por ejemplo, en el conjunto de datos utilizado en la sección anterior.
* Se puede esperar que cuando un bibliotecario busque un registro, pueda ingresar el identificador único (valores en la columna Identificador) de un libro.

In [6]:
df['Identifier'].is_unique

True

Reemplacemos el índice existente con esta columna usando
* set_index

In [7]:
df = df.set_index('Identifier')
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

Podemos acceder a cada registro de forma sencilla con: ```loc[] ```

* Aunque ```loc[] ``` puede no tener un nombre tan intuitivo.
* Nos permite realizar una indexación basada en etiquetas.
* Que es el etiquetado de una fila o registro sin tener en cuenta su posición.
* En otras palabras, 206 es la primera etiqueta del índice.
* Para acceder a él por posición, podríamos usar ```df.iloc[0]```.
* Que realiza una indexación basada en la posición.

In [8]:
df.loc[206]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

##Ordenar campos

Hasta ahora, hemos eliminado columnas innecesarias y hemos cambiado el índice de nuestro ```DataFrame``` a algo más sensato.

* En esta sección, limpiaremos columnas específicas y las llevaremos a un formato uniforme para comprender mejor el conjunto de datos y reforzar la coherencia.
* En particular, limpiaremos la fecha de publicación y el lugar de publicación.

* Tras la inspección, todos los tipos de datos son actualmente el tipo de objeto, que es más o menos análogo a ```str``` en Python nativo.

* Encapsula cualquier campo que no se pueda ajustar claramente como datos numéricos o categóricos.
* Esto tiene sentido ya que estamos trabajando con datos que inicialmente son un montón de cadenas desordenadas.

In [9]:
df.dtypes.value_counts()

object    6
dtype: int64

Un campo donde tiene sentido aplicar un valor numérico es la fecha de publicación para que podamos hacer cálculos en el futuro.

In [10]:
df.loc[1905:, 'Date of Publication'].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

Un libro en particular sólo puede tener una fecha de publicación. Por lo tanto, debemos hacer lo siguiente:

* Elimine las fechas adicionales entre corchetes, siempre que estén presentes: 1879 $[1878]$.
* Convertir rangos de fechas a su "fecha de inicio", siempre que esté presente: 1860-63; 1839, 38-54.
* Elimine por completo las fechas de las que no estamos seguros y reemplácelas con NaN de NumPy: [1897?].
* Convierta la cadena nan al valor NaN de NumPy

* Sintetizando estos patrones, podemos aprovechar una única expresión regular para extraer el año de publicación.

Sintetizando estos patrones, podemos aprovechar una única expresión regular para extraer el año de publicación:

```regex = r'^(\d{4})'```

La expresión regular anterior está destinada a encontrar cuatro dígitos al comienzo de una cadena, lo cual es suficiente para nuestro caso.
* Lo anterior es una cadena sin formato.
* Lo que significa que una barra invertida ya no es un carácter de escape.
* Lo cual es una práctica estándar con expresiones regulares.

* ```\d``` representa cualquier dígito y ```{4}``` repite esta regla cuatro veces.
* El carácter ```^``` coincide con el inicio de una cadena.
* Los paréntesis indican un grupo de captura, que indica a Pandas que queremos extraer esa parte de la expresión regular.
* Queremos que ```^``` evite casos en los que ```[ ``` comience con la cadena ```)```.


Si no estás familiarizado con las expresiones regulares, puedes inspeccionar la expresión anterior en "[regex101.com](https://regex101.com/)" y aprender todo sobre las expresiones regulares.

In [11]:
extr = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

Técnicamente, esta columna todavía tiene tipo de objeto, pero podemos obtener fácilmente su versión numérica con ```pd.to_numeric```.

In [12]:
df['Date of Publication'] = pd.to_numeric(extr)
df['Date of Publication'].dtype

dtype('float64')

Esto da como resultado que falte aproximadamente uno de cada diez valores, lo cual es un pequeño precio a pagar por poder hacer cálculos sobre los valores válidos restantes.


In [13]:
df['Date of Publication'].isnull().sum() / len(df)

0.11717147339205986

##Combinando métodos

Combinando métodos ```str``` con NumPy para limpiar columnas.

* Anteriomente utilizamos: ```df['Fecha de publicación'].str```.
* Este atributo es una forma de acceder a operaciones rápidas de cadenas en Pandas.
* Que imitan en gran medida operaciones en cadenas nativas de Python.
* O expresiones regulares compiladas, como ```.split()```, ```.replace()``` y ```.capitalize()```.

* Para limpiar el campo ```Lugar de publicación```.
* Podemos combinar los métodos pandas ```str``` con la función ```np.where``` de NumPy.
* Que es básicamente una forma vectorizada de la macro ```IF()``` de Excel.
* Tiene la forma de: ```np.where(condition, then, else)```.


In [14]:
df['Place of Publication'].head(10)

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
Name: Place of Publication, dtype: object

Vemos que en algunas filas el lugar de publicación está rodeado de otra información innecesaria.
* Si miráramos más valores, veríamos que este es el caso solo para algunas filas cuyo lugar de publicación es "Londres" u "Oxford".

In [15]:
df.loc[4157862]

Place of Publication                                  Newcastle-upon-Tyne
Date of Publication                                                1867.0
Publisher                                                      T. Fordyce
Title                   Local Records; or, Historical Register of rema...
Author                      FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4157862, dtype: object

In [16]:
df.loc[4159587]

Place of Publication                                  Newcastle upon Tyne
Date of Publication                                                1834.0
Publisher                                                Mackenzie & Dent
Title                   An historical, topographical and descriptive v...
Author                                              Mackenzie, E. (Eneas)
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4159587, dtype: object

Estos dos libros se publicaron en el mismo lugar, pero uno tiene guiones en el nombre del lugar mientras que el otro no.

* Para limpiar esta columna de un solo barrido, podemos usar ```str.contains()``` para obtener una máscara booleana.

In [17]:
pub = df['Place of Publication']
london = pub.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [18]:
oxford = pub.str.contains('Oxford')

Aquí, la función ```np.where``` se llama en una estructura anidada, siendo la condición una serie de booleanos obtenidos con ```str.contains()```.
* El método ```contains()``` funciona de manera similar a la palabra clave incorporada utilizada para encontrar la aparición de una entidad en un iterable (o subcadena en una cadena).

* El reemplazo a utilizar es una cadena que representa nuestro lugar de publicación deseado.
* También reemplazamos guiones con un espacio con ```str.replace()``` y los reasignamos a la columna en nuestro ```DataFrame```.

* Aunque hay más datos sucios en este conjunto de datos, por ahora solo discutiremos estas dos columnas.


In [19]:
df['Place of Publication'] = np.where(london, 'London',
                                      np.where(oxford, 'Oxford',
                                               pub.str.replace('-', ' ')))

In [20]:
df['Place of Publication'].head()

Identifier
206    London
216    London
218    London
472    London
480    London
Name: Place of Publication, dtype: object

In [21]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...